src/decision/dMCDA.jl

```diff
@@ -737,87 +721,111 @@ function guided_site_selection(
-
-    idx_to_replace = sort(unique(reinterpret(Int64, pref_dists)))
-    select_n = length(idx_to_replace)
-
-    keep_idxs = setdiff(collect(1:length(pref_locs)), idx_to_replace)
-
-    # Store of new set of locations
-    updated_locs = pref_locs
+    area_to_seed::Float64,
+    available_space::Vector{Float64},
+    n_site_int::Int64,
+    n_reefs::Int64,
+)
+    # Get full ordering of locations
+    loc_order = s_order[:, 1]
+    # Get unique reefs in location set
+    unique_reefs = unique(reefs)
+    pref_locs = loc_order[1:n_site_int]
+
+    num_locs = n_site_int # No. of locations to consider
+
+    for ll in 1:length(loc_order)
+        # If enough space for seeding corals, keep n_site_int, else expand as needed
+        cumulative_space = cumsum(available_space[loc_order])
+        num_locs = if cumulative_space[n_site_int] .< area_to_seed
+            findfirst(>=(area_to_seed), cumulative_space)
+        else
+            n_site_int
+        end
 
-    while (length(alt_sites) .>= select_n)
-        updated_locs = [updated_locs[keep_idxs[:]]; alt_sites[1:select_n]]
+        pref_locs = loc_order[1:num_locs]
+        pref_reefs = reefs[pref_locs] # Reefs that selected locations sit within
 
-        # Find all sites within these highly ranked but unselected sites which are further apart
-        alt_dists = dist[updated_locs, updated_locs] .> min_dist
+        # Number of times a selected location appears within each reef
+        sum_pref_locs = dropdims(
+            sum(pref_reefs .== reshape(unique_reefs, 1, length(unique_reefs)); dims=1);
+            dims=1,
+        )
 
-        # Select from these sites those far enough away from all sites
-        keep_idxs = sum(alt_dists, dims=2) .== n_sites - 1
+        # If more than n_reefs locations in a reef, swap out the worst locations
+        reefs_swap = unique_reefs[(sum_pref_locs .> n_reefs)]
+
+        if !isempty(reefs_swap)
+            # Locations to replace
+            replace_locs = [
+                pref_locs[pref_reefs .== gp][(n_reefs + 1):end] for gp in reefs_swap
+            ]
+            replace_locs = replace_locs[1:end...]
+            reef_add = unique_reefs[(sum_pref_locs .+ 1) .<= n_reefs] # acceptable reefs to swtich out for
+
+            pref_locs = setdiff(pref_locs, replace_locs) # Remove locations to be replaced from preferred sites
+            loc_order = setdiff(loc_order, replace_locs) # Remove locations to be replaced from site order
+            add_locs = setdiff(loc_order, pref_locs) # Locations which can be added in place
+
+            # Indices of locations available that satisfy the reef constraint
+            add_locs_ind = findall(
+                dropdims(
+                    any(
+                        reshape(reefs[add_locs], 1, length(reefs[add_locs])) .== reef_add;
+                        dims=1,
+                    );
+                    dims=1,
+                ),
+            )
+            # New preferred location set
```

Comment:
Add space for readability

```suggestion

            # New preferred location set
```